[Boost] Add admin notification bubble on cloud css failure

dilirity opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes https://github.com/Automattic/boost-cloud/issues/160.

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* This PR adds a notification bubble next to Boost if there's a problem with Cloud CSS.

![image](https://user-images.githubusercontent.com/11799079/207003514-04b02d7c-f111-401e-9563-b7dbbbcad9af.png)

#### Other information:

For the time being, the code only checks if something's wrong with Cloud CSS. It should be easy to update it to include another Boost feature if we need to (like say the Image Guide).

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
N/A

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No.

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Enable "**Automatically Optimize CSS Loading**" from Boost's dashboard;
* Setup your test website to use your local Boost Cloud;
* Break the Cloud CSS generation;
* Make a request to regenerate CSS through the WP-admin;
* Once you get the error, refresh the page;
* You should see the number 1 in red (dark orange?) next to Boost in the sidebar.

dilirity committed:
Add get_state_status method to Critical CSS State

dilirity committed:
Add menu notification bubble if there's a problem with Cloud CSS

dilirity committed:
changelog

haqadn commented on projects/plugins/boost/app/lib/Features_Status.php:
```diff
@@ -0,0 +1,35 @@
+	}
+
+	private static function get_cloud_css_state_status() {
+		$is_enabled = '1' === get_option( 'jetpack_boost_status_critical-css' );
```
Though this is precisely what we need, there is an abstraction we use for module statuses, and I think we should continue to use them rather than checking the option directly.
This would be the recommended way:

```
use Automattic\Jetpack_Boost\Features\Optimizations\Critical_CSS\Critical_CSS;
use Automattic\Jetpack_Boost\Lib\Status;

$status = new Status( Critical_CSS::get_slug() );
if( $status->is_enabled() ) { ... }
```

dilirity replied:
Thank you! I was looking for a way to do it using the available API, but couldn't find it.

Fixed with 08499227213e2db9d865e1de077ac2552230c1a5.

haqadn commented on projects/plugins/boost/app/lib/Features_Status.php:
```diff
@@ -0,0 +1,35 @@
+ */
+class Features_Status {
+
+	const GOOD = 'good';
+	const BAD  = 'bad';
```
If the only purpose of `good` and `bad` are to identify if an error exists or not, the flag could be just a boolean. If we think in terms of `has_cloud_css_problem` instead of `get_cloud_css_state_status`, a boolean would be sufficient. I think that would also simplify the class and make it more intuitive.

dilirity replied:
Fixed with 08499227213e2db9d865e1de077ac2552230c1a5.

I guess I was thinking around the idea that a feature has three states - good, bad and inactive. I then decided to combine inactive and good into one, but I didn't re-think the whole logic.

Also - great suggestion with the method name! That made the code far easier to read.

haqadn commented on projects/plugins/boost/app/lib/critical-css/Critical_CSS_State.php:
```diff
@@ -160,6 +160,15 @@ public function set_as_failed( $error ) {
+	 *
+	 * @return mixed
+	 */
+	public function get_state_status() {
```
It's probably just me and let me know if you disagree. Since the class is already `Critical_CSS_State` perhaps we could just call `get_status`. I guess, having a "state" property in a class named "state" made things already pretty confusing. 🤪

dilirity replied:
Oops, you're absolutely right. I was bouncing between classes and totally missed, that there's already state in there haha.

dilirity replied:
Fixed with 55b3c63e0c33e61570596b3c18bed03c6fb180e2.

I guess we have quite a few methods with `state` in their name 😅

haqadn commented on projects/plugins/boost/app/lib/Features_Status.php:
```diff
@@ -0,0 +1,35 @@
+	const BAD  = 'bad';
+
+	public static function get_total_problems() {
+		$status = self::get_cloud_css_state_status();
+		return $status === self::GOOD ? 0 : 1;
```
@thingalon suggested this, and I think it would be elegant as well. Rather than manually checking everything within this method, we could use a filter to get the count of problems. This would look something like this:

```
return apply_filter( 'jetpack_boost_total_problem_count', 0 );
```

And then, each active optimization feature or anything else can hook into this and increase the count. This has the advantage of letting the features worry about determining how they identify problems.

dilirity replied:
I'm having a bit of trouble with this one. The `apply_filters` code would get executed way before features are initialised:
https://github.com/Automattic/jetpack/blob/4d5f0b4814f6d0d3e48402b5e3492a2e583d8fc7/projects/plugins/boost/app/admin/class-admin.php#L62-L73
I added a health check method to [Cloud_CSS](https://github.com/Automattic/jetpack/blob/4d5f0b4814f6d0d3e48402b5e3492a2e583d8fc7/projects/plugins/boost/app/features/optimizations/cloud-css/Cloud_CSS.php), but I'm not sure where to hook it so it gets initialised before the menu does.

I imagine adding the bubble with JS is out of the question 😅

thingalon replied:
> I'm having a bit of trouble with this one. The apply_filters code would get executed way before features are initialised

🤔 Oh. That does make it annoying.

`Admin_Menu::add_menu` has to be called before the `admin_menu` hook. All of our optimizations get initialized by the `init` hook.

We could move this call to `Admin_Menu::add_menu` to a late-firing `init` hook, or an early-firing `admin_menu` hook.

dilirity replied:
Thank you for the pointers @thingalon!

I've made some changes to the way checks are handled here 5fb2c30dad44ce4fb4006e92ca55324b1c782f5e.

Those include moving the `Admin_Menu::add_menu` to the `admin_menu` hook. I went through the code and there shouldn't be any problems with this move.

I also scrapped the whole Features_Status class in favour of methods inside the feature itself. Though I'm not a 100% happy with having to hook the method from within each feature. I couldn't find the right place to add a loop that goes through all features, checks if they have the "health check" method and updating the problem count if necessary.

haqadn reviewed (dismissed):
Nice work! This does what we need very well. I have suggested some improvements. Let me know if you need any clarification.

haqadn commented:
Btw, some tests appear to be failing. Please take a look into that as well.

dilirity committed:
Update method name

dilirity committed:
Update features status to use abstraction for module statuses

dilirity committed:
Remove feature status in favor of features checking their own status

This would make it easier for features to have their own logic for
checking when something's wrong. It would also be easier to just hook
on the filter and update the problem count.

pyronaur commented on projects/plugins/boost/app/features/optimizations/cloud-css/Cloud_CSS.php:
```diff
@@ -213,4 +214,28 @@ public function add_critical_css_constants( $constants ) {
+	 * @param  int $count The current problem count.
+	 * @return int
+	 */
+	public function update_total_problem_count( $count ) {
```
Could this method be missing a check whether $count is an `int`?

dilirity replied:
That's a good point - thanks!

thingalon reviewed (approved):
Looks good! 🚢
